Feature/1535 anonymise test server data

richard-jones opened the pull request:
This PR is not quite ready to be merged, but this is a place to review the current (nearly complete) state.

Here's the additions to the anonymisation code to handle export and import in convenient ways.

When reviewing this you also need to look at this bit in esprit, which does the grunt work:

https://github.com/CottageLabs/esprit/blob/master/esprit/tasks.py#L9

To do the export you can just run

```
python portality/scripts/anon_export.py
```

This will place the output in `local_store/main/anon_export`

To reimport, you can run

```
python portality/scripts/anon_import.py data_import_settings/dev_basics.json
```

This will look at the file `data_import_settings/dev_basics.json` and import a selection of the types and just 100000 of the articles.

Steven-Eardley commented on portality/scripts/anon_import.py:
```diff
@@ -0,0 +1,94 @@
+    with codecs.open(args.config, "rb", "utf-8") as f:
+        config = json.loads(f.read())
+
+    do_import(config)
```
Can ```config``` be undefined here?

richard-jones replied:
no.  config is a positional command line argument, so if it is omitted argparse will give you an error

Steven-Eardley commented on portality/scripts/anon_import.py:
```diff
@@ -0,0 +1,94 @@
+            if handle is None:
+                break
+            print("Retrieved {x} from storage".format(x=filename))
+            tempStore.store(container, filename, source_stream=handle)
```
We don't have any specific exception handling here, are we happy for the whole import to fail if storage has a blip?

richard-jones replied:
For the time being, yes.  The storage implementation is weakly tested at this stage and we decided that we'd test it properly when we want to use it for something more serious and it gets a proper implementation.  In the mean time this developer-oriented script is allowed to be a bit more fragile, I think.

Steven-Eardley commented on requirements.txt:
```diff
@@ -4,7 +4,7 @@
 # first get esprit, specifically the version we know will work (+ it's
 # not on PyPI)
--e git+https://github.com/CottageLabs/esprit.git@d91e9d52e27ebd1da0a97809a5a56e250ca6e803#egg=esprit
+-e git+https://github.com/CottageLabs/esprit.git@b8d4e5cab053b952bf1a2f6ff68f315c052ff33f#egg=esprit
```
This has been updated since this tag

emanuil-tolev commented on portality/scripts/anon_import.py:
```diff
@@ -0,0 +1,94 @@
+def do_import(config):
+    host = app.config["ELASTIC_SEARCH_HOST"]
+    index = app.config["ELASTIC_SEARCH_DB"]
+    if config.get("elastic_search_host") is not None:
```
To always have to specify the host so it makes you think about it? If the alternative is to imply "localhost" if not specified, then it's possible you accidentally start importing into localhost without meaning to.

richard-jones replied:
This allows you to specify an override to your configuration if you want, for whatever reason, to import the data to an index other than the one your app is set up to read from.  If you don't want to do that, then you never need to do anything, and you won't notice this feature.

emanuil-tolev commented:
Looks sensible and pretty robust with the new Store service.

richard-jones commented on portality/scripts/anon_import.py:
```diff
@@ -0,0 +1,94 @@
+        index = config.get("elastic_search_db")
+        app.config["ELASTIC_SEARCH_DB"] = index
+
+    print("\n")
```
Unfortunately not.  This print statement adds an extra line after the block of print statements before it.  Since print() implicitly includes a newline, if we also added it to the block before it, everything would have an extra line between them, and there wouldn't be a differentiating bit of whitespace between that block and the next.

I'm not sure this is wrong, and is just personal preference :)

richard-jones commented on portality/scripts/anon_import.py:
```diff
@@ -0,0 +1,94 @@
+
+    print("\n")
+    print("Using host {x} and index {y}\n".format(x=host, y=index))
+    conn = esprit.raw.make_connection(None, host, None, index)
```
You'd have to change esprit for that

richard-jones commented on portality/scripts/anon_import.py:
```diff
@@ -0,0 +1,94 @@
+    print("\n")
+
+    if config.get("confirm", True):
+        text = raw_input("Continue? [y/n]")
```
Done

richard-jones commented on portality/scripts/anon_import.py:
```diff
@@ -0,0 +1,94 @@
+
+    # remove all the types that we are going to import
+    for import_type in import_types.keys():
+        esprit.raw.delete(conn, import_type)
```
If there's no type there it will just fail silently

richard-jones commented on portality/scripts/anon_import.py:
```diff
@@ -0,0 +1,94 @@
+            if limit is not None and imported_count != -1:
+                limit -= imported_count
+            if limit <= 0:
+                break
```
hm, not sure why that was working, I've made this test more explicit

Steven-Eardley commented on portality/settings.py:
```diff
@@ -120,6 +120,29 @@
+# ================================
+# File store
+
+STORE_IMPL = "portality.store.StoreS3"
```
This could be dangerous to the live data - it defaults to using S3 with the live bucket. If someone accidentally uses an admin key (or one that writes) they could mess up the data.

This should be ```None``` and print a warning on start, and be set in `production.cfg` for the live server.

Steven-Eardley commented:
Now that @anusharanganathan has added print statements for the boto s3 errors, a successful import outputs this:

```
==Importing==
Importing all from account
Obtaining account from storage
Retrieved account.bulk.1 from storage
Error fetching #account.bulk.2
An error occurred (NoSuchKey) when calling the GetObject operation: The specified key does not exist.
Importing all from journal
Obtaining journal from storage
Retrieved journal.bulk.1 from storage
Error fetching #journal.bulk.2
An error occurred (NoSuchKey) when calling the GetObject operation: The specified key does not exist.
```

boto3 | 1.9.10
botocore | 1.12.11

richard-jones commented:
In the updated version I just pushed I've modified the error handling.  This now looks for NoSuchKey in the `StoreS3.get` method and returns None, which is the behaviour I was originally expecting.

I've also addressed the defaulting - this now defaults to local store, and you have to set it to use the S3 store in production.cfg

Steven-Eardley commented:
After I pulled your changes I was seeing an error caused by gzip / the stream from s3:

I don't know whether it's because the stored objects aren't gzipped yet, or whether the stream handle is incompatible with ```.tell()```

```
==Importing==
Importing all from account
Obtaining account from storage
Retrieved account.bulk.1 from storage
Traceback (most recent call last):
  File "portality/scripts/anon_import.py", line 96, in <module>
    do_import(config)
  File "portality/scripts/anon_import.py", line 67, in do_import
    tempStore.store(container, filename, source_stream=ziphandle)
  File "/Users/steve/code/cl/doaj/src/doaj/portality/store.py", line 133, in store
    f.write(source_stream.read())
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/gzip.py", line 261, in read
    self._read(readsize)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/gzip.py", line 295, in _read
    pos = self.fileobj.tell()   # Save current position
AttributeError: 'StreamingBody' object has no attribute 'tell'
```

richard-jones commented:
The problem with unzipping the stream from s3 has been fixed (well, bypassed), and this now has gone through a cycle of testing on the test server, and synchronising the exported data to my local machine.  I think this means it is now ready.  Any final reviews please on this, then I'll merge it and we'll run on live asap.
